lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -1,94 +1,477 @@
-"""
-Helper functions to get the elasticsearch specific configurations
-"""
-
+import requests
+from datetime import datetime
+from dateutil import parser as date_parser
 from dateutil import rrule
+from dateutil.relativedelta import relativedelta
+from enum import Enum
+from flask import request
+from flask_restful import Resource, abort
+from logging import Logger
+from typing import Any, AnyStr, Callable, Dict, List
+from urllib.parse import urljoin
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.auth import Auth
 
-from configparser import NoSectionError, NoOptionError
 
+class SchemaError(TypeError):
+    """
+    Generic base class for errors in processing a JSON schema.
+    """
 
-def get_user_term(user: str):
-    """Elasticsearch doesn't index (or search for) null values, and it's not
-    clear we really want to consider "owner": null to be "owned" datasets, in
-    any case. (Nor will they be possible beyond the transition period where we
-    may retain limited support for 0.69 data.)
+    pass
 
-    Instead, let a query for data with a missing or null "user" parameter find
-    all data that's set to allow public access. This by default includes all
-    "unowned" data, but also data owned by a user that's been published.
+
+class InvalidRequestPayload(SchemaError):
+    """
+    A required client JSON input document is missing.
     """
-    if user:
-        term = {"authorization.owner": user}
-    else:
-        term = {"authorization.access": "public"}
-    return term
 
+    def __str__(self):
+        return "Invalid request payload"
 
-def get_es_host(config):
-    try:
-        return config.get("elasticsearch", "host")
-    except (NoOptionError, NoSectionError):
-        return ""
 
+class MissingParameters(SchemaError):
+    """
+    One or more required JSON keys are missing, or the values are unexpectedly
+    empty.
+    """
 
-def get_es_port(config):
-    try:
-        return config.get("elasticsearch", "port")
-    except (NoOptionError, NoSectionError):
-        return ""
+    def __init__(self, keys: List[AnyStr]):
+        self.keys = keys
+
+    def __str__(self):
+        return f"Missing required parameters: {','.join(self.keys)}"
+
+
+class ConversionError(SchemaError):
+    """
+    ConversionError Used to report an invalid parameter type
+    """
+
+    def __init__(self, value: Any, expected_type: str, actual_type: str):
+        """
+        __init__ Construct a ConversionError exception
+
+        Args:
+            value: The value we tried to convert
+            expected_type: The expected type
+            actual_type: The actual type
+        """
+        self.value = value
+        self.expected_type = expected_type
+        self.actual_type = actual_type
 
+    def __str__(self):
+        return f"Value {self.value} ({self.actual_type}) cannot be parsed as a {self.expected_type}"
```

Comment:
It could be; I'm not sure it's necessarily more helpful, but I'll repr it anyway.

Reply:
Oh, good:  you used the `!r` directive rather than my actual suggestion -- that's what I _should_ have suggested, but it didn't occur to me until today.